File: luogu/p3387.cpp
#include <iostream>
#include <algorithm>
#include <queue>
#include <stack>
#include <cstring>
const int N = 50009;
using namespace std;
struct Edge {
    int u;
    int v;
    int next;
}e[N], edge[N];
typedef pair<int, int> pii;
stack<int> q;
queue<int> p;
int dfn[N], low[N], h[N], w[N], d[N], fa[N], head[N], in[N];
int dis[N];
int instack[N];
int s, edgenum, dnum, ans;
void addEdge(int u, int v) {
    edgenum++;
    e[edgenum].u = u;
    e[edgenum].v = v;
    e[edgenum].next = h[u];
    h[u] = edgenum;
}
void tarjan(int u)
{
    s++;
    low[u] = s;
    dfn[u] = s;
    q.push(u);
    instack[u] = 1;
    for (int i = h[u]; i; i = e[i].next) {
        int v = e[i].v;
        if (!dfn[v]) tarjan(v);
        if (instack[v]) low[u] = min(low[u], low[v]);
    }
    if (low[u] == dfn[u]) {
        int flag = 0, cnt = 0, od = 0;
        while (1) {
            int v = q.top();
            q.pop();
            instack[v] = 0;
            if (v == u) break;
            w[u] += w[v];
            w[v] = 0;
            fa[v] = u;
        }
        d[++dnum] = u;
    }
}
void topo() {
    
    for (int i = 1; i <= dnum; i++) if (in[d[i]] == 0) { 
        p.push(d[i]); dis[d[i]] = w[d[i]]; 
    }
    while (!p.empty()) {
        int u = p.front();
        p.pop();
        for (int i = head[u]; i; i = edge[i].next) {
            int v = edge[i].v;
            dis[v] = max(dis[v], dis[u] + w[v]);
            in[v]--;
            if (in[v] == 0) p.push(v);
        }
    }
    for (int i = 1; i <= dnum; i++) ans = max(ans, dis[d[i]]);
}

int main()
{
    int n, m;
    cin >> n >> m;
    for (int i = 1; i <= n; i++) cin >> w[i];
    for (int i = 1; i <= m; i++) {
        int u, v;
        cin >> u >> v;
        addEdge(u, v);
    }
    for (int i = 1; i <= n; i++) fa[i] = i;
    for (int i = 1; i <= n; i++) if (!dfn[i]) tarjan(i);
    edgenum = 0;
    for (int i = 1; i <= m; i++) {
        int u = fa[e[i].u], v = fa[e[i].v];
        if (u != v) {
            edgenum++;
            edge[edgenum].u = u;
            edge[edgenum].v = v;
            edge[edgenum].next = head[u];
            head[u] = edgenum;
            in[v]++;
        }

    }
    topo();
    cout << ans << endl;
    return 0;
}

File: luogu/p3387.py
import collections
import fileinput
class Graph:
    def __init__(self,size=0)->None:
        self.head=[0]*size
        self.edge=[[0,0,0]]
        self.w=[0]
        self.size=size
        self.inDegree=[0]*size
    def addEdge(self,edge)->None:
        u,v=edge[0],edge[1]
        self.edge.append([u,v,self.head[u]])
        self.head[u]=len(self.edge)-1
    def addValue(self,value:list)->None:
        self.w+=value
    def topo(self)->int:
        vis=collections.deque()
        dis=self.w.copy()
        for i in range(1,self.size):
            if self.inDegree[i]==0:
                vis.append(i)
        while vis:
            u=vis.popleft()
            i=self.head[u]
            while i:
                v=self.edge[i][1]
                dis[v]=max(dis[v],dis[u]+self.w[v])
                self.inDegree[v]-=1
                if self.inDegree[v]==0:
                    vis.append(v)
                i=self.edge[i][2]
        return max(dis)        

class Spn:
    def __init__(self,graph:Graph)->None:
        self.dfncnt=0
        self.low=[0]*graph.size
        self.dfn=[0]*graph.size
        self.vis=collections.deque()
        self.fa=list(range(graph.size))
        self.w=[]
        self.dic={}
        self.graph=graph
    def tarjan(self,u:int)->None:
        self.dfncnt+=1
        self.low[u]=self.dfncnt
        self.dfn[u]=self.dfncnt
        self.vis.append(u)
        i=self.graph.head[u]
        while i:
            v=self.graph.edge[i][1]
            if (self.dfn[v]==0):
                self.tarjan(v)
            if v in self.vis:
                self.low[u]=min(self.low[u],self.low[v])
            i=self.graph.edge[i][2]
        if (self.low[u]==self.dfn[u]):
            while True:
                v=self.vis.pop()
                if (u==v): 
                    break
                self.graph.w[u]+=self.graph.w[v]
                self.graph.w[v]=0
                self.fa[v]=u
            self.w.append(self.graph.w[u])
            self.dic[u]=len(self.w)
    def shrink(self,graph:Graph)->Graph:
        for i in range(1,graph.size):
            if self.dfn[i]==0:
                self.tarjan(i)
        g=Graph(len(self.w)+1)
        for i in graph.edge:
            u,v=self.fa[i[0]],self.fa[i[1]]
            if (u!=v):
                u,v=self.dic[u],self.dic[v]
                g.addEdge([u,v])
                g.inDegree[v]+=1
        g.addValue(self.w) 
        return g


            
if __name__=="__main__":
    file=open('D:\\code\\python\\luogu\\input.txt','r')
    n,m=map(int,file.readline().split())
    g=Graph(n+1)
    g.addValue(list(map(int,file.readline().split())))
    for i in range(m):
        g.addEdge(list(map(int,file.readline().split())))
    spn=Spn(g)
    g=spn.shrink(g)
    print(g.topo())
    file.close()
    


